File: 1520-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/number-of-steps-to-reduce-a-number-in-binary-representation-to-one.cpp
class Solution {
private:
    void addOne(string &s){
        int carry=0;
        for(int i=s.length()-1; i>=0; i--){
            if(s[i]=='1'){
                s[i]='0';
                carry=1;
            }else{
                s[i]='1';
                carry=0;
                break;
            }
        }
        if(carry) s= '1' + s;
    }
public:
    int numSteps(string s) {
        int count=0;
        while(s!="1"){
            // check for odd or even
            if(s.back()=='0'){
                // means even hai
                s.pop_back(); // right shift basically
            }else addOne(s);
            count++;
        }
        return count;
    }
};

File: 1520-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/number-of-steps-to-reduce-a-number-in-binary-representation-to-one.py
class Solution:
    def numSteps(self, s: str) -> int:
        s=list(s)
        def addOne(binary):
            carry=0
            n=len(binary)
            for i in range(n-1,-1,-1):
                if binary[i]=="1":
                    carry=1
                    binary[i]="0"
                else:
                    if carry==0:
                        break
                    else:
                        binary[i]="1"
                        carry=0
                        break
            # if carry is one, add extra 1 to the left
            if carry==1:
                binary.insert(0,"1")
        count=0 
        while True:
            if s==["1"]:
                return count
            count+=1
                # either odd or even
            if s[-1]=="0":
                #means it is even
                s.pop()
                # print(s)
            else:
                addOne(s)
        return count 
                    
                    
            
            
        